firedrake/assemble_expressions.py

```diff
@@ -358,17 +386,36 @@ def pointwise_expression_kernel(exprs, scalar_type):
                            remove_zeros=False, emit_return_accumulate=False)
     coefficients = translator.varmapping
     args = []
+    plargs = []
     for expr in exprs:
-        retval = coefficients.pop(expr.lvalue)
-        args.append(loopy.GlobalArg(retval.name, shape=retval.shape, dtype=expr.lvalue.dat.dtype))
-        for c in expr.rcoefficients:
-            if c.dat == expr.lvalue.dat:
+        for c, arg in zip(expr.coefficients, expr.args):
+            try:
+                var = coefficients.pop(c)
+            except KeyError:
                 continue
-            var = coefficients[c]
+            plargs.append(arg)
             args.append(loopy.GlobalArg(var.name, shape=var.shape, dtype=c.dat.dtype))
+    assert len(coefficients) == 0
```

Comment:
Merging of arguments across the grouped expressions (basically constants only come in once)